pinot-core/src/main/java/org/apache/pinot/core/plan/AggregationPlanNode.java

```diff
@@ -62,31 +68,164 @@ public AggregationPlanNode(IndexSegment indexSegment, QueryContext queryContext)
   public Operator<IntermediateResultsBlock> run() {
     assert _queryContext.getAggregationFunctions() != null;
 
+    boolean hasFilteredPredicates = _queryContext.isHasFilteredAggregations();
+    if (hasFilteredPredicates) {
+      return buildFilteredAggOperator();
+    }
+
+    return buildNonFilteredAggOperator();
+  }
+
+  /**
+   * Returns {@code true} if the given aggregations can be solved with segment metadata, {@code false} otherwise.
+   * <p>Aggregations supported: COUNT
+   */
+  private static boolean isFitForMetadataBasedPlan(AggregationFunction[] aggregationFunctions) {
+    for (AggregationFunction aggregationFunction : aggregationFunctions) {
+      if (aggregationFunction.getType() != AggregationFunctionType.COUNT) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Returns {@code true} if the given aggregations can be solved with dictionary, {@code false} otherwise.
+   * <p>Aggregations supported: MIN, MAX, MIN_MAX_RANGE, DISTINCT_COUNT, SEGMENT_PARTITIONED_DISTINCT_COUNT
+   */
+  private static boolean isFitForDictionaryBasedPlan(AggregationFunction[] aggregationFunctions,
+      IndexSegment indexSegment) {
+    for (AggregationFunction aggregationFunction : aggregationFunctions) {
+      AggregationFunctionType functionType = aggregationFunction.getType();
+      if (functionType != AggregationFunctionType.MIN && functionType != AggregationFunctionType.MAX
+          && functionType != AggregationFunctionType.MINMAXRANGE
+          && functionType != AggregationFunctionType.DISTINCTCOUNT
+          && functionType != AggregationFunctionType.SEGMENTPARTITIONEDDISTINCTCOUNT) {
+        return false;
+      }
+      ExpressionContext argument = (ExpressionContext) aggregationFunction.getInputExpressions().get(0);
+      if (argument.getType() != ExpressionContext.Type.IDENTIFIER) {
+        return false;
+      }
+      String column = argument.getIdentifier();
+      Dictionary dictionary = indexSegment.getDataSource(column).getDictionary();
+      if (dictionary == null) {
+        return false;
+      }
+    }
+    return true;
+  }
```

Comment:
Can we move these 2 methods back? They are not changed, but make code review and future tracking harder